homeassistant/components/econet/water_heater.py

```diff
@@ -1,236 +1,162 @@
-        for key, value in self.econet_state_to_ha.items():
-            self.ha_state_to_econet[value] = key
-        for mode in self.supported_modes:
-            if mode not in ECONET_STATE_TO_HA:
-                error = f"Invalid operation mode mapping. {mode} doesn't map. Please report this."
-                _LOGGER.error(error)
-
-    @property
-    def name(self):
-        """Return the device name."""
-        return self.water_heater.name
-
-    @property
-    def available(self):
-        """Return if the the device is online or not."""
-        return self.water_heater.is_connected
 
-    @property
-    def temperature_unit(self):
-        """Return the unit of measurement."""
-        return TEMP_FAHRENHEIT
+    def on_update_received(self):
+        """Update was pushed from the ecoent API."""
+        if self._running != self.water_heater.running:
+            # Water heater running state has changed so check usage on next update
+            self._poll = True
+            self._running = self.water_heater.running
+        super().on_update_received()
 
     @property
     def device_state_attributes(self):
         """Return the optional device state attributes."""
-        data = {}
-        vacations = self.water_heater.get_vacations()
-        if vacations:
-            data[ATTR_VACATION_START] = vacations[0].start_date
-            data[ATTR_VACATION_END] = vacations[0].end_date
-        data[ATTR_ON_VACATION] = self.water_heater.is_on_vacation
-        todays_usage = self.water_heater.total_usage_for_today
-        if todays_usage:
-            data[ATTR_TODAYS_ENERGY_USAGE] = todays_usage
-        data[ATTR_IN_USE] = self.water_heater.in_use
-
-        if self.water_heater.lower_temp is not None:
-            data[ATTR_LOWER_TEMP] = round(self.water_heater.lower_temp, 2)
-        if self.water_heater.upper_temp is not None:
-            data[ATTR_UPPER_TEMP] = round(self.water_heater.upper_temp, 2)
-        if self.water_heater.is_enabled is not None:
-            data[ATTR_IS_ENABLED] = self.water_heater.is_enabled
-
-        return data
+        _attr = super().device_state_attributes
+        _attr[ATTR_SUPPORTS_LEAK] = self.water_heater.leak_installed
+        if self.water_heater.tank_hot_water_availability:
+            _attr[
+                ATTR_TANK_HOT_WATER_AVAILABILITY
+            ] = self.water_heater.tank_hot_water_availability
+        if self.water_heater.has_shutoff_valve:
+            _attr[ATTR_SHUT_OFF_VALVE_CLOSED] = not self.water_heater.shutoff_valve_open
+        if self.water_heater.tank_health:
+            _attr[ATTR_TANK_HEALTH] = self.water_heater.tank_health
+        if self.water_heater.override_status:
+            _attr[ATTR_OVERRIDE_STATUS] = self.water_heater.override_status
+        if self.water_heater.todays_energy_usage is not None:
+            _attr[ATTR_ENERGY_USAGE] = round(self.water_heater.todays_energy_usage, 2)
```

Comment:
Measurements that are relevant on their own without the water heater state should be moved to separate sensor entities. Since we're making breaking changes, it's a good time to make the integration follow our current design standards.

Reply:
energy used and water used are updated based on the water heater's state changing from running to not running. Does that mean they shouldn't be moved?

What about hot water availability that is directly related to the water heater running/user changing the temp.

Sounds like maybe shutoff valve status, tank health, and compressor health should maybe be moved?

Is there anything I need to know to get these sensors to link to the device?

Reply:
Energy and water used are relevant measurements on their own. To give an example, the brightness of a light is not relevant without knowing if the light is on or off.

What does hot water availability measure?

Yes, shutoff valve status, tank health, and compressor health, sounds like they should be moved.

To link the entities to the device, set the correct identifiers in the `device_info` entity property:
https://developers.home-assistant.io/docs/device_registry_index#defining-devices

Reply:
This makes sense. Hot water available is just that, a percentage of available hot water. It should be moved as well based on the above. I'll work on this. Thanks.

Reply:
@MartinHjelmare moved all the attributes to sensors.

One issue I had was that they don't get updated unless I add polling. This is because the pyeconet library only allows one callback to be passed to the econet object for when it receives updates. Any idea on how I can fix that? Could update the library to accept multiple call backs... but it would be nice if I can tell HA to force a check of all the sensors when the water heater gets an update.

Reply:
The callback should be handled centrally in the integration and then we use our dispatch helper to signal interested entities in their respective platforms.

https://github.com/home-assistant/core/blob/62237adf91e542ce389163917b62e0ced1cf3f50/homeassistant/helpers/dispatcher.py#L32-L37

https://github.com/home-assistant/core/blob/62237adf91e542ce389163917b62e0ced1cf3f50/homeassistant/helpers/dispatcher.py#L77-L80